### ISSUE TITLE
[dynamo] Some inefficiencies around handling __torch_function__

### ISSUE THREAD

**Issue by anijain2305** [pytorch/pytorch#151776]
### 🐛 Describe the bug

I was looking at reducing compile time for a GGUF SD model (https://github.com/pytorch/pytorch/issues/150706), and I found some inefficiencies around `__torch_function__`.  The model heavily relies on torch function. 

Testing on a single transformer layer, I was able to reduce Dynamo time from 3 seconds to below 2 seconds with this patch. I am not confident that this patch is good, it will be good if someone can take over. 

I tried to reduce the number of times `__torch_function__` was called. I saw that for torch.compile, it was getting called 3x more times. A majority of these calls were not getting traced by Dynamo, which hinted that they were either coming from compiler themselves (like calling x._is_view to check something can call `__torch_function__`). 

```
diff --git a/torch/_dynamo/output_graph.py b/torch/_dynamo/output_graph.py
index 2dc2209a2c6..6228ec0d949 100644
--- a/torch/_dynamo/output_graph.py
+++ b/torch/_dynamo/output_graph.py
@@ -1471,7 +1471,8 @@ class OutputGraph:
                 self.tracing_context.fake_mode = backend_fake_mode

             with self.restore_global_state():
-                compiled_fn = self.call_user_compiler(gm)
+                with torch._C.DisableTorchFunction():
+                    compiled_fn = self.call_user_compiler(gm)

             from torch.fx._lazy_graph_module import _LazyGraphModule

diff --git a/torch/_dynamo/polyfills/tensor.py b/torch/_dynamo/polyfills/tensor.py
index 002ccf5d1d4..b3d81036ab3 100644
--- a/torch/_dynamo/polyfills/tensor.py
+++ b/torch/_dynamo/polyfills/tensor.py
@@ -11,25 +11,26 @@ from ..decorators import substitute_in_graph
 def make_subclass(
     cls: type[Any], data: torch.Tensor, requires_grad: bool = False, **kwargs: Any
 ) -> Any:
-    # This is a rough approximation of `THPVariable_make_subclass`. It should
-    # suffice for most of Dynamo tracing purposes.
-    # https://github.com/pytorch/pytorch/blob/ccfde4dadfa3c342076a1ee387017f84dd4ad2f7/torch/csrc/autograd/python_variable.cpp#L597-L650
-    assert len(kwargs) == 0, "_make_subclass only supports requires_grad as keyword arg"
-    data = data.detach()
-
-    # Avoid unnecessary `requires_grad` mutation, which isn't supported in Dynamo.
-    if data.requires_grad != requires_grad:
-        data.requires_grad = requires_grad
-
-    # Dynamo can't yet handle upcasting to base tensor type via `as_subclass`.
-    if cls is torch.Tensor:
-        return torch.Tensor(data)
-
-    # Calling `as_subclass` because
-    # 1. Dynamo knows how to handle it
-    # 2. the C impls match at this point -- both `THPVariable_make_subclass` and
-    #    `THPVariable_as_subclass` calls `THPVariable_NewWithVar`.
-    return data.as_subclass(cls)
+    with torch._C.DisableTorchFunctionSubclass():
+        # This is a rough approximation of `THPVariable_make_subclass`. It should
+        # suffice for most of Dynamo tracing purposes.
+        # https://github.com/pytorch/pytorch/blob/ccfde4dadfa3c342076a1ee387017f84dd4ad2f7/torch/csrc/autograd/python_variable.cpp#L597-L650
+        assert len(kwargs) == 0, "_make_subclass only supports requires_grad as keyword arg"
+        data = data.detach()
+
+        # Avoid unnecessary `requires_grad` mutation, which isn't supported in Dynamo.
+        if data.requires_grad != requires_grad:
+            data.requires_grad = requires_grad
+
+        # Dynamo can't yet handle upcasting to base tensor type via `as_subclass`.
+        if cls is torch.Tensor:
+            return torch.Tensor(data)
+
+        # Calling `as_subclass` because
+        # 1. Dynamo knows how to handle it
+        # 2. the C impls match at this point -- both `THPVariable_make_subclass` and
+        #    `THPVariable_as_subclass` calls `THPVariable_NewWithVar`.
+        return data.as_subclass(cls)


 __all__ = [
diff --git a/torch/_dynamo/variables/builder.py b/torch/_dynamo/variables/builder.py
index 6e80e1ef563..b05f2403f56 100644
--- a/torch/_dynamo/variables/builder.py
+++ b/torch/_dynamo/variables/builder.py
@@ -425,7 +425,8 @@ class VariableBuilder:
         if cached_vt:
             return cached_vt

-        vt = self._wrap(value)
+        with torch._C.DisableTorchFunctionSubclass():
+            vt = self._wrap(value)
         vt.source = self.source
         if (
             self._can_lift_attrs_to_inputs(vt)
diff --git a/torch/_dynamo/variables/torch_function.py b/torch/_dynamo/variables/torch_function.py
index dd7f6fa1f53..c376e2a745f 100644
--- a/torch/_dynamo/variables/torch_function.py
+++ b/torch/_dynamo/variables/torch_function.py
@@ -628,7 +628,7 @@ class TensorWithTFOverrideVariable(TensorVariable):

         # Handle non-overriden attributes inherited from `torch.Tensor`.
         attr_is_overriden = _is_attr_overidden(tx, self, name)
-        if hasattr(torch.Tensor, name) and not attr_is_overriden:
+        if hasattr(torch.Tensor, name) and not attr_is_overriden and not inspect.ismethoddescriptor(getattr(torch.Tensor, name)):
             if tx.output.torch_function_enabled:
                 if self.source:
                     install_guard(
```

Changing my "eager" backend to do this might have helped as well

```

def my_backend(gm, *args):
    def inner(*n_args):
        with torch._C.DisableTorchFunction():
            return gm.forward(*n_args)
    return inner
```

cc @StrongerXi @mlazos 

### Versions

NA

**Comment by StrongerXi** [comment]
I tried this patch (both as is, and applying individual `with torch._C.DisableTorchFunctionSubclass()` only) on the e2e repro in #150706, but I didn't really observe any difference compilation time.

**Comment by StrongerXi** [comment]
I did more analysis on compilation time of the original FP16 model (no `torch_function` or subclass) v.s. Q2K model (with `torch_function` subclass). The later also has the patch from #151265, to remove some `detach` nodes and around 10s Dynamo time (tlparse [here](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpnpYS9R/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000)).

- [fp16 tlparse](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpfsXenl/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000), Dynamo ~15s, 2318 graph ops.
- [q2k tlparse](https://manifold.edge.x2p.facebook.net/v0/read/tree/logs/.tmpeYTuzS/index.html?bucketName=tlparse_reports&apiKey=tlparse_reports-key&withPayload=1&timeoutMsec=10000), Dynamo ~104s, 12478 graph ops.

Looking at the Dynamo graphs, I noticed that the q2k version basically has these extra bits (dequantization before matmul):
```verbatim
        # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:416 in __torch_function__, code: result = super().__torch_function__(func, types, args, kwargs)
        result_20: "u8[18432, 1008][1008, 1]cuda:0" = l_self_modules_joint_transformer_blocks_modules_0_modules_norm1_context_modules_linear_parameters_weight_.view(torch.uint8);  l_self_modules_joint_transformer_blocks_modules_0_modules_norm1_context_modules_linear_parameters_weight_ = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:416 in __torch_function__, code: result = super().__torch_function__(func, types, args, kwargs)
        result_21: "u8[221184, 84][84, 1]cuda:0" = result_20.reshape((221184, 84));  result_20 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:416 in __torch_function__, code: result = super().__torch_function__(func, types, args, kwargs)
        split_1 = torch.functional.split(result_21, [16, 64, 2, 2], dim = 1);  result_21 = None
        x_5: "u8[221184, 16][84, 1]cuda:0" = split_1[0]
        x_6: "u8[221184, 64][84, 1]cuda:0" = split_1[1]
        x_7: "u8[221184, 2][84, 1]cuda:0" = split_1[2]
        x_8: "u8[221184, 2][84, 1]cuda:0" = split_1[3];  split_1 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:416 in __torch_function__, code: result = super().__torch_function__(func, types, args, kwargs)
        result_22: "f16[221184, 1][42, 1]cuda:0" = x_7.view(torch.float16);  x_7 = None
        result_23: "f16[221184, 1][42, 1]cuda:0" = result_22.to(None);  result_22 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:416 in __torch_function__, code: result = super().__torch_function__(func, types, args, kwargs)
        result_24: "f16[221184, 1][42, 1]cuda:0" = x_8.view(torch.float16);  x_8 = None
        result_25: "f16[221184, 1][42, 1]cuda:0" = result_24.to(None);  result_24 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:416 in __torch_function__, code: result = super().__torch_function__(func, types, args, kwargs)
        result_26: "u8[221184, 16][16, 1]cuda:0" = x_5.__and__(15)
        result_27: "f16[221184, 16][16, 1]cuda:0" = result_23.mul(result_26);  result_23 = result_26 = None
        result_28: "f16[221184, 16, 1][16, 1, 1]cuda:0" = result_27.reshape((221184, 16, 1));  result_27 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:416 in __torch_function__, code: result = super().__torch_function__(func, types, args, kwargs)
        result_29: "u8[221184, 16][16, 1]cuda:0" = x_5.__rshift__(4);  x_5 = None
        result_30: "f16[221184, 16][16, 1]cuda:0" = result_25.mul(result_29);  result_25 = result_29 = None
        result_31: "f16[221184, 16, 1][16, 1, 1]cuda:0" = result_30.reshape((221184, 16, 1));  result_30 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:343 in dequantize_blocks_Q2_K, code: shift = torch.tensor([0, 2, 4, 6], device=d.device, dtype=torch.uint8).reshape((1, 1, 4, 1))
        tensor_2: "u8[4][1]cuda:0" = torch.tensor([0, 2, 4, 6], device = device(type='cuda', index=0), dtype = torch.uint8)
        shift_1: "u8[1, 1, 4, 1][4, 4, 1, 1]cuda:0" = tensor_2.reshape((1, 1, 4, 1));  tensor_2 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:416 in __torch_function__, code: result = super().__torch_function__(func, types, args, kwargs)
        result_32: "u8[221184, 2, 1, 32][84, 32, 32, 1]cuda:0" = x_6.reshape((221184, -1, 1, 32));  x_6 = None
        result_33: "u8[221184, 2, 4, 32][256, 128, 32, 1]cuda:0" = result_32.__rshift__(shift_1);  result_32 = shift_1 = None
        result_34: "u8[221184, 2, 4, 32][256, 128, 32, 1]cuda:0" = result_33.__and__(3);  result_33 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:416 in __torch_function__, code: result = super().__torch_function__(func, types, args, kwargs)
        result_35: "u8[221184, 16, 16][256, 16, 1]cuda:0" = result_34.reshape((221184, 16, 16));  result_34 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:416 in __torch_function__, code: result = super().__torch_function__(func, types, args, kwargs)
        result_36: "f16[221184, 16, 16][256, 16, 1]cuda:0" = result_28.mul(result_35);  result_28 = result_35 = None
        result_37: "f16[221184, 16, 16][256, 16, 1]cuda:0" = result_36.sub(result_31);  result_36 = result_31 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:416 in __torch_function__, code: result = super().__torch_function__(func, types, args, kwargs)
        result_38: "f16[221184, 256][256, 1]cuda:0" = result_37.reshape((221184, -1));  result_37 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:416 in __torch_function__, code: result = super().__torch_function__(func, types, args, kwargs)
        result_39: "f16[18432, 3072][3072, 1]cuda:0" = result_38.reshape((18432, 3072));  result_38 = None
        
         # File: /home/ryanguo99/pt/pytorch/torch/_dynamo/polyfills/tensor.py:27 in make_subclass, code: return torch.Tensor(data)
        weight_2: "f16[18432, 3072][3072, 1]cuda:0" = torch.Tensor(result_39);  result_39 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:454 in forward, code: weight = weight.to(self.compute_dtype)
        weight_3: "bf16[18432, 3072][3072, 1]cuda:0" = weight_2.to(torch.bfloat16);  weight_2 = None
        
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/quantizers/gguf/utils.py:457 in forward, code: output = torch.nn.functional.linear(inputs, weight, bias)
        output_1: "bf16[2, 18432][18432, 1]cuda:0" = torch._C._nn.linear(silu_2, weight_3, None);  silu_2 = weight_3 = None
```

Whereas the fp16 version just has the `nn.linear` call, without all the dequantization bits:
```verbatim
         # File: /home/ryanguo99/.conda/envs/comfyui/lib/python3.12/site-packages/diffusers/models/normalization.py:168 in forward, code: emb = self.linear(self.silu(emb))
        silu_1: "bf16[2, 3072][3072, 1]cuda:0" = torch.nn.functional.silu(sample_2, inplace = False)
        emb_5: "bf16[2, 18432][18432, 1]cuda:0" = torch._C._nn.linear(silu_1, l_self_modules_joint_transformer_blocks_modules_0_modules_norm1_modules_linear_parameters_weight_, None);  silu_1 = l_self_modules_joint_transformer_blocks_modules_0_modules_norm1_modules_linear_parameters_weight_ = None
```

These `torch._C._nn.linear` calls repeat 326 times, and each time the q2k graph has 29 more ops, resulting in 9454 more ops in total, accounting for 93% of the extra ops seen in the q2k graph.

Not sure how much we can optimize Dynamo here (all the `torch_function` emulation overhead), that feels like a project on its own. So hierarchical compilation might be the best tool here? @anijain2305
